solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,429 @@
+ * <li>Documents (of a particular type)</li>
+ * </ul>
+ * 
+ * When instantiating this DataContext, an index name is provided. Within this
+ * index, each document type is represented as a table.
+ * 
+ * This implementation supports either automatic discovery of a schema or manual
+ * specification of a schema, through the {@link SimpleTableDef} class.
+ */
+
+public class SolrDataContext extends QueryPostprocessDataContext implements
+		DataContext {
+
+	private static final Logger logger = LoggerFactory
+			.getLogger(SolrDataContext.class);
+
+	public static final String FIELD_ID = "_id";
+
+	private final SimpleTableDef tableDef;
+	private final String indexName;
+	private final String url;
+
+	private SolrQuery query = new SolrQuery();
+
+	private static enum queryTypes {
+		ROW, FACET
+	};
+
+	private static final int MAX_RETRIES = 0;
+	private static final int SOCK_TIMEOUT = 1000;
+	private static final int CONN_TIMEOUT = 5000;
+	private static final int DEFAULT_LIMIT = 1000;
+
+	public SolrDataContext(String url, String indexName) {
+		this.url = url;
+		this.indexName = indexName;
+		this.tableDef = detectSchema(url, indexName);
+	}
+
+	private SimpleTableDef detectSchema(String url, String indexName) {
+		SimpleTableDef tableDef = null;
+
+		url += "/schema";
+
+		try {
+			HttpClient httpclient = new DefaultHttpClient();
+			HttpGet request = new HttpGet(url);
+			HttpResponse response = httpclient.execute(request);
+
+			InputStreamReader inputStream = new InputStreamReader(response
+					.getEntity().getContent());
+
+			ObjectReader reader = new ObjectMapper().reader(HashMap.class);
+			Map<String, Object> outerMap = reader.readValue(inputStream);
+
+			Map<String, Object> schemaMap = (Map) outerMap.get("schema");
+
+			if (outerMap != null) {
+				List<Map<String, String>> fieldMapList = (List) schemaMap
+						.get("fields");
+				List<String> columnNames = new ArrayList<String>();
+				List<ColumnType> columnTypes = new ArrayList<ColumnType>();
+
+				for (int i = 0; i < fieldMapList.size(); i++) {
+					Map<String, String> fieldMap = fieldMapList.get(i);
+
+					columnNames.add(fieldMap.get("name"));
+					columnTypes.add(ColumnType.VARCHAR);
+				}
+
+				String[] columnNamesArr = new String[columnNames.size()];
+				columnNamesArr = columnNames.toArray(columnNamesArr);
+
+				ColumnType[] columnTypesArr = new ColumnType[columnTypes.size()];
+				columnTypesArr = columnTypes.toArray(columnTypesArr);
+
+				tableDef = new SimpleTableDef(indexName, columnNamesArr,
+						columnTypesArr);
+			}
+		} catch (IOException e) {
+			logger.error("Failed to parse schema", "", e);
+		}
+
+		return tableDef;
+	}
+
+	@Override
+	protected Schema getMainSchema() throws MetaModelException {
+		try {
+			final MutableSchema theSchema = new MutableSchema(
+					getMainSchemaName());
+
+			final MutableTable table = tableDef.toTable().setSchema(theSchema);
+			final Column idColumn = table.getColumnByName(FIELD_ID);
```

Comment:
Does a field named '_id' always exist in Solr? Or else I would expect some special treatment to introduce this as a virtual/surrogate field in the ```detectSchema``` method above.